ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/SupportsMempool.hs

```diff
@@ -166,3 +153,124 @@ type GenTxId blk = TxId (GenTx blk)
 class HasTxs blk where
   -- | Return the transactions part of the given block in no particular order.
   extractTxs :: blk -> [GenTx blk]
+
+{-------------------------------------------------------------------------------
+  Tx sizes
+-------------------------------------------------------------------------------}
+
+-- | Each block has its limits of how many transactions it can hold. That limit
+-- is compared against the sum of measurements taken of each of the
+-- transactions in that block.
+--
+-- How we measure the transaction depends of the era that this transaction
+-- belongs to (more specifically it depends on the block type to which this
+-- transaction will be added). For initial eras (like Byron and initial
+-- generations of Shelley based eras) this measure was simply a byte size
+-- (block could not be bigger then given size - in bytes - specified by the
+-- ledger state). In future eras (starting with Alonzo) this measure was a bit
+-- more complex as it had to take other factors into account (like execution
+-- units). For details please see the individual instances for the TxLimits.
+class ( Measure     (TxMeasure blk)
+      , HasByteSize (TxMeasure blk)
+      , NoThunks    (TxMeasure blk)
+      , Show        (TxMeasure blk)
+      ) => TxLimits blk where
+  -- | The (possibly multi-dimensional) size of a transaction in a block.
+  type TxMeasure blk
+
+  -- | The various sizes (bytes, Plutus script ExUnits, etc) of a tx /when it's
+  -- in a block/
+  --
+  -- This size is used to compute how many transaction we can put in a block
+  -- when forging one.
+  --
+  -- The byte size component in particular might differ from the size of the
+  -- serialisation used to send and receive the transaction across the network.
+  -- For example, CBOR-in-CBOR could be used when sending the transaction
+  -- across the network, requiring a few extra bytes compared to the actual
+  -- in-block serialisation. Another example is the transaction of the
+  -- hard-fork combinator which will include an envelope indicating its era
+  -- when sent across the network. However, when embedded in the respective
+  -- era's block, there is no need for such envelope. An example from upstream
+  -- is that the Cardano ledger's "Segregated Witness" encoding scheme
+  -- contributes to the encoding overhead.
+  --
+  -- INVARIANT Assuming no hash collisions, the size should be the same in any
+  -- state in which the transaction is valid.
```

Comment:
Add reference scripts as an example where the condition "in any state in which the transaction is valid" becomes relevant?